src/libtcod/bresenham.h

```diff
@@ -95,38 +97,86 @@ namespace tcod {
     This class is provisional.
  */
 class TCODLIB_API BresenhamLine {
+  using Point2 = std::array<int, 2>;
+
  public:
   /**
       Initializes the object to draw a line from `xFrom`, `yFrom` to `xTo`, `yTo`.
   */
-  BresenhamLine(int xFrom, int yFrom, int xTo, int yTo) { TCOD_line_init_mt(xFrom, yFrom, xTo, yTo, &data_); }
+  BresenhamLine(Point2 from, Point2 to) {
+    TCOD_bresenham_data_t data;
+    TCOD_line_init_mt(from.at(0), from.at(1), to.at(0), to.at(1), &data);
+    do {
+      data_.push_back(from);
+    } while (!TCOD_line_step_mt(&from.at(0), &from.at(1), &data));
+  }
 
   /**
       Steps through each cell from the start to the end coordinates.
 
       The input variables `x`, `y` are passed in by reference and updated at each step of the line.
   */
-  inline bool step(int& x, int& y) { return TCOD_line_step_mt(&x, &y, &data_); }
+  inline bool step(int& x, int& y) {
+    auto pos = data_.at(cur_);
+    x = pos.at(0);
+    y = pos.at(1);
+    return (cur_++ < data_.size());
+  }
+
+  struct iterator : public std::iterator<std::bidirectional_iterator_tag, Point2> {
+    iterator(std::size_t cur, std::vector<value_type>& data) : cur_(cur), data_(data) {}
+
+    inline iterator operator++(int) {
+      auto tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+    inline iterator& operator++() {
+      ++cur_;
+      return *this;
+    }
+    inline iterator& operator--() {
+      --cur_;
+      return *this;
+    }
+
+    inline reference operator*() const { return data_.at(cur_); }
+    inline bool operator==(const iterator& rhs) const { return (data_ == rhs.data_ && cur_ == rhs.cur_); }
+    inline bool operator!=(const iterator& rhs) const { return !(*this == rhs); }
+
+   private:
+    std::size_t cur_;
+    std::vector<value_type>& data_;
+  };
+
+  iterator begin() { return iterator(0, data_); }
+  iterator end() { return iterator(data_.size(), data_); }
 
  private:
-  TCOD_bresenham_data_t data_{};
+  std::size_t cur_ = 0;
+  std::vector<Point2> data_{};
 };
 
 /**
     Draw a Bresenham line, passing the indexes of the line to `callback`.
 
     This function is provisional.
  */
-inline bool bresenham_line(int xFrom, int yFrom, int xTo, int yTo, const std::function<bool(int, int)>& callback) {
-  auto line = BresenhamLine(xFrom, yFrom, xTo, yTo);
-  do {
-    if (!callback(xFrom, yFrom)) {
+inline bool bresenham_line(
+    std::array<int, 2> from, std::array<int, 2> to, const std::function<bool(std::array<int, 2>)>& callback) {
+  for (auto&& [x, y] : BresenhamLine(from, to)) {
+    if (!callback({x, y})) {
```

Comment:
While I mean to target C++17 and support structured bindings the libtcod library itself targets C++11 and is tested with C++14.  This means you can't use structured bindings in libtcod, only in your own C++17 code which uses libtcod.
```suggestion
  for (auto&& xy : BresenhamLine(from, to)) {
    if (!callback(xy)) {
```

Reply:
I saw the warnings about C++17 too, my bad. Easy enough to fix up though.